niceman/utils.py

```diff
@@ -895,4 +896,21 @@ def unicode(s):
     else:
         return __builtin__.unicode(s)
 
+
+def generate_unique_name(pattern, nameset):
+    """Create a unique numbered name from a pattern and a set"""
+    i = 0
+    while True:
+        n = pattern % i
+        i += 1
+        if n not in nameset:
+            return n
+
+
+# http://stackoverflow.com/questions/1151658/python-hashable-dicts
+class HashableDict(dict):
+    """Dict that can be used as keys"""
+    def __hash__(self):
+        return hash(frozenset(self.values()))
```

Comment:
OK, you've sufficiently twisted my arm.  I've been staring at the YAML configuration file, considering it to be the goal, and viewing it as a nice recursive dict/array structure with no methods. So I've been avoiding defining these method-less classes to maximize flexibility and avoiding a proliferation of code. But, creating hash-able objects is the first in-my-face pragmatic advantage of defining and using one of these classes to store data. 

Once I clean up the other issues you have identified, I will create the Origin object (with a child AptOrigin object). This will have a nice side effect in that I should rewrite the code to fully define the AptOrigin list as the Packages are identified (instead of doing the two passes that I do now). If you like how it looks, then I'll proceed defining other objects (like Package/DpkgPackage).